cheri_compressed_cap_common.h

```diff
@@ -67,22 +67,32 @@ _CC_STATIC_ASSERT_SAME(_CC_N(FIELD_INTERNAL_EXPONENT_SIZE) + _CC_N(FIELD_EXP_NON
                            _CC_N(FIELD_EXPONENT_LOW_PART_SIZE),
                        _CC_N(FIELD_EBT_SIZE));
 
+// In order to allow vector loads and store from memory we can optionally reverse the first two fields.
 struct _cc_N(cap) {
     /* offset = cursor - base */
+#ifdef _CC_REVERSE_PESBT_CURSOR_ORDER
+    /* Currently only updated lazily when compress is called.
+     * We might want to make this the canonical field and slice out other fields on demand.*/
+    uint64_t cached_pesbt;
     _cc_addr_t _cr_cursor; /* Capability cursor */
-    _cc_addr_t cr_base;    /* Capability base addr */
+#else
+    _cc_addr_t _cr_cursor;
+    _cc_addr_t cached_pesbt;
+#endif
     _cc_length_t _cr_top;  /* Capability top */
+    _cc_addr_t cr_base;    /* Capability base addr */
     uint32_t cr_perms;     /* Permissions */
     uint32_t cr_uperms;    /* User Permissions */
     uint32_t cr_otype;     /* Object Type, 24/16 bits */
     /* Original EBT from the decompressed capability. If you modify
      * cursor/base/top, you must be sure to either recalculate this field to
      * match or that the result is still representable, since this will be the
      * EBT written out to memory */
-    uint32_t cr_ebt;     /* Exponent/Base/Top */
     uint8_t cr_flags;    /* Flags */
     uint8_t cr_reserved; /* Remaining hardware-reserved bits to preserve */
     uint8_t cr_tag;      /* Tag */
+    uint8_t cr_bounds_valid /* Set if bounds decode was given an invalid cap */;
+    uint64_t padding;
```

Comment:
Still don't like that this exists

Reply:
Given that TCG is breaking outside of the type system, and I am relying heavily on the exact size of the struct, I think explicit padding is better.

Reply:
But that's where offsetof and sizeof come in?

Reply:
Is the problem that there is explicit padding, or that I have calculated it manually? I could always add a wrapping struct, but then that causes lots of mess everywhere to do cap->inner.foo rather than cap->foo. I don't know how to do sizeof with an anonymous member.

Reply:
Explicit padding. Why would you need a wrapping struct?

Reply:
A wrapping struct is what I would need to avoid manually calculating the padding. As in 
`struct outer{
  struct inner{...};
  char pad[TARGET_SIZE - sizeof(struct inner)];
}`.
Which is nice in that it is less manual, but creates slightly messier code elsewhere. I like explict padding, because I require the struct to be a certain size for some optimised TCG elsewhere, and I don't want to rely on knowledge of when the compiler will/wont insert padding.

Reply:
Yes, but your problem is that you are adding explicit padding whether manually or automatically calculated. It's also unlikely to remain true on all architectures (either 32-bit or the notoriously-weird m68k so it's still _wrong_ even if we don't care about them). Where in TCG do you care how many bytes this struct is? Can that not just use sizeof?

Reply:
I can use sizeof to automatically disable the optimisation (or, rather, throw an assert so I can go fix the padding for that target). If you want to go have a look where this is used, its in the QEMU PR, called "gen_vector_copy". I was trying to use vector tcg types for capability operations to avoid massive amounts of tcg ops to move decompressed structs around. Every capability instruction has to do such a move (to move from source to dest). If the struct is 512 bits, this is 4 TCG ops on a target with TCG_TARGET_HAS_v256 defined. If the the struct were just a byte shorter, then you would need 12 TCG ops. Obviously, the compiler inserts some padding, so one byte less will probably never happen. But rather than rely on the compiler for padding here, given I explicitly want it, I think its best I add it.

Reply:
If you want to pad to a multiple of a specific size then you have two options:
1. Mark the struct as aligned, though then you waste space for temporaries on the stack, and _make this configurable_ since non-QEMU doesn't need QEMU-specific gunk like that
2. Mark the fields in QEMU that need to be as aligned as aligned
This approach is wrong IMO.

Reply:
So I have specified the alignment requirement in QEMU in the GPCapRegs struct (because that is the context in which it needs to be aligned). Can I also declare an array of QEMU_ALIGNED(64) T and have the compiler insert padding automatically? Or can alignment only be specified when defining T?

Reply:
No, you can only align the start of the array, otherwise pointer arithmetic would break. 2 would require a wrapper for any array elements. Though that's probably not a bad thing IMO, gives you full control over what padding you want inside QEMU?

Reply:
I suppose if I do that with an anonymous member then it won't create the code mess I was worried about.

Reply:
Ah, but then I would need to be declaring the cap_register_t inside my aligned thing.

Reply:
Guess I will just deal with X->cap.foo instead of X->foo everywhere =(